src/Model/Mail.php

```diff
@@ -18,6 +19,70 @@
  */
 class Mail
 {
+	/**
+	 * Insert received private message
+	 *
+	 * @param array $msg
+	 * @return int|boolean Message ID or false on error
+	 */
+	public static function insert($msg)
+	{
+		$user = User::getById($msg['uid']);
+
+		if (empty($msg['convid'])) {
+			$mail = DBA::selectFirst('mail', ['convid'], ["`convid` != 0 AND `parent-uri` = ?", $msg['parent-uri']]);
+			if (DBA::isResult($mail)) {
+				$msg['convid'] = $mail['convid'];
+			}
+		}
+
+		if (empty($msg['guid'])) {
+			$host = parse_url($msg['from-url'], PHP_URL_HOST);
+			$msg['guid'] = Item::guidFromUri($msg['uri'], $host);
+		}
+
+		DBA::lock('mail');
```

Comment:
Please use `Lock::accquire()`, the `DBA::lock()` shouldn't be used directly, it's a lock driver.

Reply:
No, this is a direct lock of the database table "mail". This is a database mechanism to block the "mail" table from being modified by another process.

Reply:
The locking mechanism should prevent such situations. I don't feel well that we use different locking mechanism for similar tasks. What if another process locks with `Lock::acquire ('mail')` and it get locked through Redis/Memcache. So the lock would return true, but the same lock would return false in case another process blocked it per `DBA::lock()`. We shouldn't rely on just DB specific processes I think.

Reply:
These are two completely different things. The one is some semaphore locking process which is meant to avoid parallel execution of a specific task. The other one is a database mechanism to block specific tables from being written by any process, anywhere in the system. These locks will never interfere.

Reply:
I know what you mean, but isn't a semaphore locking better for it? The goal is: If there are two processes, which creates the same mail, don't add the second one? This should be the same as:

```php
if (Lock::acquire('mail') {
    if (DBA::exists(...)) {
        Lock::release();
        return;
    }

    DBA::insert(...);
    Lock::release();
}
```

Which else process should insert **exactly** the same mail (uri, uid) except this process? And why would we like to block the whole table for it? I think we should avoid DB-locks. I know it from business that really bad things could happen in such cases ... I don't think that this happen to you, but if somebody else adds `DBA::insert('mail', ...)` out of other scenarios ..

We could go a little bit further:
```php
if (Lock::acquire('mail|' . $msg['uri'] . '|' . $msg['uid']) {
   ...
}
```
Now we could send mails parallel without synchronization the DB in the backend

Reply:
Acquiring massively locks like you described in the last lines can be a huge problem when the semaphore based locking is used. This is limited to a small amount of parallel locks.

And: This here really is the textbook example of when to use a database lock. We want to lock exactly that table, because we want to write in it. This is not comparable to other places like in the worker handling, a database lock on the workerqueue table would not work.

The mail table isn't written very often and the locked time is short. So there is no problem.